### DISCUSSION TITLE
Scans of the daughter board

### DISCUSSION THREAD

**Discussion by KaruroChori** [xblax/flashforge_ad5m_klipper_mod#155] [Ideas]
Are there scans of the daughter board on the tool-head already available online?
I made a quick search, but without a positive outcome.
Worst case scenario I could take some photos.

**Comment by consp** [comment]
> Are there scans of the daughter board on the tool-head already available online?

I've seen them but do not remember where, I think it was on reddit somewhere, might be a good idea to make them.

**Comment by KaruroChori** [comment]
![QSCF0064](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9377545/89c88ab7-5205-4d39-b1db-a3d82b2fde2f)
Not exactly great but I had no easy way to fit external lights and macro lens in there. Still it should be readable.

**Reply by KaruroChori** [reply]
I see the unpopulated pads for a ~~JTAG~~ SWD programmer right on top.
In theory it is possible to connect a programmer, dump the flash memory and check the parameters used to compile klipper from that, right?

In case it was not clear my attempt was to bring klipper from master on this printer as well :D.

**Reply by consp** [reply]
> In case it was not clear my attempt was to bring klipper from master on this printer as well :D.

You've gotten hold of the Nations SDK? The analogue pheripherals are different from the STM variant afaik or isn't that a problem?

**Reply by KaruroChori** [reply]
To be honest I just assumed it was an STM32 clone of sort, and I could have used the same cheap clone programmer to handle basic functionality :D to recover the flashed image.
I see the n32g455 in the list of supported devices by klipper 0.12, so in theory as long as I can compile and image and flash it there it should be fine? Assuming klipper properly implemented the hal for this target.

**Reply by consp** [reply]
If I remember correctly someone tried it and got nowhere except when compiling for the stm32, but maybe you have more luck.

**Reply by KaruroChori** [reply]
Ok, I guess I will start by dumping the image from there as long as there is no permanent fuse set to prevent that and my stlink clone works.
Worst case scenario, I should be able to recover it with that.

**Reply by xblax** [reply]
I think that these MCUs use non-standard bootloader, so be careful with what you do. @g992 did a lot of experiments on running klipper master in the past, but had problems to make it work.

**Reply by g992** [reply]
I have dump of eboard, yes, it have custom bootloader, if u need, i can share .bin file (cant attach to message).
If u compile klipper as is top n32 (in stm32 section), mcu just dont booting (:
If u compile for stm32f103, mvu boot ok, but analog perepherials not working.
i just resolder n32 to stm32f103 from bluepill, compile to f103, and all works good

**Reply by KaruroChori** [reply]
Yes, please, if you can share the bin file it would really help!

**Reply by g992** [reply]
https://pixeldrain.com/u/X3VDPtNU - .bin file
https://pixeldrain.com/u/dN3EhRWN -.zip with full chip dump (!) 400M .bin inside archive

**Reply by KaruroChori** [reply]
Thank you!

**Reply by Tasshack** [reply]
> I have dump of eboard, yes, it have custom bootloader, if u need, i can share .bin file (cant attach to message). If u compile klipper as is top n32 (in stm32 section), mcu just dont booting (: If u compile for stm32f103, mvu boot ok, but analog perepherials not working. i just resolder n32 to stm32f103 from bluepill, compile to f103, and all works good

Does the stock firmware works with that after replacing the firmware?

**Comment by KaruroChori** [comment]
For reference, this is the stock configuration as self reported:
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9377545/cadf85dc-f288-4cf6-b35e-b00e6dbfb243)

**Reply by KaruroChori** [reply]
Also for reference, the official SDK
https://github.com/NationsTechCoreLib/N32G455xx/tree/main

**Comment by The-Synthax** [comment]
There is an external oscillator on the backside of the toolhead board. What is its frequency?

**Reply by KaruroChori** [reply]
That photo was never taken but the board self reports running at 144MHz (image above)

**Reply by The-Synthax** [reply]
> That photo was never taken but the board self reports running at 144MHz (image above)

The reference clock will be a different frequency, so the text on this oscillator is important for getting this working correctly. The datasheet reports a maximum external clock frequency of 32MHz.

**Reply by KaruroChori** [reply]
Got it! I am unable to check it for now, and it would probably have to wait next week. If anyone else can, please do that :).

**Reply by The-Synthax** [reply]
> Got it! I am unable to check it for now, and it would probably have to wait next week. If anyone else can, please do that :).

Okay! I don't have a 5M, but I will soon. I maintain a Klipper port for Adventurer 3 on my repo. Don't worry, we'll get this working soon enough, and I plan to add support for adding an RPi to supplement the Allwinner. Hopefully GuppyScreen or even KlipperScreen will work without issue afterward, but will have to see how much RAM those end up taking up.

**Comment by alankessler** [comment]
There are front and back photos here: https://www.machines-3d.com/en/ad5mad5mpro-extruder-board-xml-856_935_859-6891.html

Full resolution here:
https://www.machines-3d.com/images/Image/EXTRU.png
and 
https://www.machines-3d.com/images/Image/EXTRU1.png
Can you see what you need?

**Reply by KaruroChori** [reply]
@The-Synthax, just in case they got no notification for this.

**Reply by The-Synthax** [reply]
> There are front and back photos here: https://www.machines-3d.com/en/ad5mad5mpro-extruder-board-xml-856_935_859-6891.html
> 
> Full resolution here: https://www.machines-3d.com/images/Image/EXTRU.png and https://www.machines-3d.com/images/Image/EXTRU1.png Can you see what you need?

Perfect! Exactly what I need. Thank you!

**Reply by MrNRod** [reply]
> > There are front and back photos here: https://www.machines-3d.com/en/ad5mad5mpro-extruder-board-xml-856_935_859-6891.html
> > Full resolution here: https://www.machines-3d.com/images/Image/EXTRU.png and https://www.machines-3d.com/images/Image/EXTRU1.png Can you see what you need?
> 
> Perfect! Exactly what I need. Thank you!

Just so I am clear, is the work you are doing essentially reverse engineering the Daugher Board so we can complie klipper from source for these boards? 

I was interested in potential replacing the Allwinner T113-s3 with the -s4 variant since it has twice the built in ram. And unless they hardcoded a memory limit it should work as a drop in. Since I'm more compelled to keep the board mostly stock and don't want to have an external device running klipper.

**Reply by The-Synthax** [reply]
@MrNRod This should be fine, but you probably need to compile your own kernel for OpenWRT with the s4 in mind and swap it in for the existing one. The rootfs shouldn’t need any changes, as far as I am aware. 

Currently I’m only gathering the info I need to get upstream Klipper to run on the tool head _period_, but as soon as I have a 5M in front of me I’ll get to work figuring out the best way to flash it (with future ease of updates in mind.) I shouldn’t need to do any additional reverse engineering as I believe all of its config/pin table is already available from FlashForge. Should be pretty quick once I get my hands on the thing, ASSUMING time/life/responsibilities permit.

**Comment by The-Synthax** [comment]
Can anyone point me to where in the init process Klipper is uploaded to the MCU? It's probably someplace incredibly obvious but I seem to be overlooking where this happens.

Edit: Wait... Are we even running upstream Klipper on the mainboard? Or just using FlashForge's?

**Reply by cwestpha** [reply]
To my understanding, you don't write klipper to the MCU or toolhead PCB outside of firmware updates. Just like mainline. The custom bootloads they use are likely just so they can do end to end updating of the components for firmware updates without having dedicated debug lines for completing this task (like a katapault or software triggered DFU).
Normal init process is SOC communicates to MCU and toolhead MCU using normal klipper comms. Then there is likely a trigger (like a GPIO or something) that triggers flash/reflash mode for firmware updates or they have another entry point to trigger programming). _The mods are just laying on top of the SoC, they dont (to my knowledge) modify anything MCU or tool head PCB side._

**Reply by The-Synthax** [reply]
> To my understanding, you don't write klipper to the MCU or toolhead PCB outside of firmware updates. Just like mainline. The custom bootloads they use are likely just so they can do end to end updating of the components for firmware updates without having dedicated debug lines for completing this task (like a katapault or software triggered DFU). Normal init process is SOC communicates to MCU and toolhead MCU using normal klipper comms. Then there is likely a trigger (like a GPIO or something) that triggers flash/reflash mode for firmware updates or they have another entry point to trigger programming). _The mods are just laying on top of the SoC, they dont (to my knowledge) modify anything MCU or tool head PCB side._

Generally no, you wouldn’t be writing anything to the MCU outside of an update. That being said, I hadn’t realized that this project doesn’t even install open source firmware on the main board MCU at all. Guess that’ll be something my project does so that we can use the latest version of Klipper and keep things up-to-date.

**Reply by KaruroChori** [reply]
Klipper, the python part being run on the SoC based on linux is upstream older version.
Klipper the C firmware installed in the MCUs is just what flashforge provided, for which the build configuration and possibly sourcecode modifications have not been shared.

**Reply by The-Synthax** [reply]
> Klipper, the python part being run on the SoC based on linux is upstream older version. Klipper the C firmware installed in the MCUs is just what flashforge provided.

Yeah I’m understanding that. I only noticed this by reading all the init and install scripts and digging through the filesystem and not finding a Klipper binary for the main board MCU. 

What got me confused was the mention of the toolhead MCU being the reason we can’t update / have new features, but sounds like that’s not accurate. It’s BOTH MCUs. I’ll get to work on that soon, upstream would be preferred.

**Reply by KaruroChori** [reply]
Yes there are two MCUs:
one on the mainboard @ https://github.com/g992/flashforge-ad5m-5mpro-research/blob/main/photo/_5m_full_view_marked.png
and the one in the toolhead which is discussed in here :). Sorry for any confusion.

**Comment by KaruroChori** [comment]
No longer needed as of https://github.com/xblax/flashforge_ad5m_klipper_mod/releases/tag/v00.06-beta
